pkg/services/ngalert/schedule/schedule_test.go

```diff
@@ -222,6 +228,78 @@ func TestAlertingTicker(t *testing.T) {
 	})
 }
 
+func TestSendingToExternalAlertmanager(t *testing.T) {
+	dbstore := tests.SetupTestEnv(t, 1)
+	t.Cleanup(registry.ClearOverrides)
+
+	// create alert rule with one second interval
+	alertRule := tests.CreateTestAlertRule(t, dbstore, 1)
+
+	fakeAM := newFakeExternalAlertmanager(t)
+	defer fakeAM.Close()
+
+	// First, let's create an admin configuration that holds an alertmanager.
+	adminConfig := &models.AdminConfiguration{OrgID: 1, Alertmanagers: []string{fakeAM.server.URL}}
+	cmd := store.UpdateAdminConfigurationCmd{AdminConfiguration: adminConfig}
+	require.NoError(t, dbstore.UpdateAdminConfiguration(cmd))
+
+	mockedClock := clock.NewMock()
+	baseInterval := time.Second
+
+	logger := log.New("ngalert schedule test")
+	schedCfg := schedule.SchedulerCfg{
+		C:                mockedClock,
+		BaseInterval:     baseInterval,
+		MaxAttempts:      1,
+		Evaluator:        eval.Evaluator{Cfg: &setting.Cfg{ExpressionsEnabled: true}, Log: logger},
+		RuleStore:        dbstore,
+		InstanceStore:    dbstore,
+		AdminConfigStore: dbstore,
+		Notifier:         &fakeNotifier{},
+		Logger:           logger,
+		Metrics:          metrics.NewMetrics(prometheus.NewRegistry()),
+	}
+	st := state.NewManager(schedCfg.Logger, nilMetrics, dbstore, dbstore)
+	sched := schedule.NewScheduler(schedCfg, nil, "http://localhost", st)
+
+	ctx := context.Background()
+
+	// Make sure we sync the configuration at least once before the evaluation happens to guarantee the sender is running
+	// when the first alert triggers.
+	require.NoError(t, sched.SyncAndApplyConfigFromDatabase())
+
+	// Then, ensure we've discovered the Alertmanager.
+	require.Eventually(t, func() bool {
+		return len(sched.AlertmanagersFor(1)) == 1 && len(sched.DroppedAlertmanagersFor(1)) == 0
+	}, 10*time.Second, 200*time.Millisecond)
+
+	go func() {
+		schedule.AdminConfigPollingInterval = 1 * time.Second
+		err := sched.Run(ctx)
+		require.NoError(t, err)
+	}()
+
+	// With everything up and running, let's advance the time to make sure we get at least one alert iteration.
+	mockedClock.Add(2 * time.Second)
+
+	// Eventually, our Alertmanager should have received at least one alert.
+	require.Eventually(t, func() bool {
+		return fakeAM.AlertsCount() >= 1 && fakeAM.AlertNamesCompare([]string{alertRule.Title})
+	}, 10*time.Second, 200*time.Millisecond)
+
+	// Now, let's remove the Alertmanager from the admin configuration.
+	adminConfig.Alertmanagers = []string{}
+	cmd = store.UpdateAdminConfigurationCmd{AdminConfiguration: adminConfig}
+	require.NoError(t, dbstore.UpdateAdminConfiguration(cmd))
+
+	// Again, make sure we sync.
+	require.NoError(t, sched.SyncAndApplyConfigFromDatabase())
```

Comment:
Since we make `AdminConfigPollingInterval` as 1s, I think we should not explicitly call it here and test if background sync is happening or not.

Reply:
I don't think this is sensible in unit tests as in general, it tends to make them flakier - if anything I think I should make the unit test NOT sync unless I tell it to e.g. set `AdminConfigPollingInterval` to `10 * time.Minute`.

In integration tests, we can test the natural flow of things as those are typically pretty long to run but for fast feedback, I don't want to up using `time.Sleep` all over since I don't have anything public to assert on.